File: app/src/main/java/com/appchamp/wordchunks/data/ChunksRealmHelper.java
package com.appchamp.wordchunks.data;

import com.appchamp.wordchunks.models.realm.Chunk;
import com.appchamp.wordchunks.models.realm.Word;
import com.appchamp.wordchunks.util.RealmUtils;
import com.appchamp.wordchunks.util.StringUtils;

import java.util.List;

import io.realm.Realm;
import io.realm.RealmList;
import io.realm.RealmResults;

import static com.appchamp.wordchunks.util.Constants.CHUNKS_SEPARATOR;
import static com.appchamp.wordchunks.util.Constants.CHUNK_STATE_NORMAL;
import static com.appchamp.wordchunks.util.Constants.REALM_FIELD_LEVEL_ID;
import static com.appchamp.wordchunks.util.Constants.REALM_FIELD_STATE;

/**
 * Chunks
 */
public final class ChunksRealmHelper {

    private ChunksRealmHelper() {
        throw new UnsupportedOperationException("u can't instantiate me...");
    }

    /**
     * Returns the list of Realm Chunk objects created from the array of split chunks.
     */
    public static RealmList<Chunk> createChunks(Realm realm, String[] wordsSplit,
                                                String levelId, RealmList<Word> wordsRealm) {
        RealmList<Chunk> chunks = new RealmList<>();
        for (int i = 0; i < wordsSplit.length; i++) {
            String[] splitChunks = StringUtils.getSplits(wordsSplit[i], CHUNKS_SEPARATOR);
            for (String chunkStr : splitChunks) {
                Chunk chunk = realm.createObject(Chunk.class);
                chunk.setChunk(chunkStr);
                chunk.setLevelId(levelId);
                chunk.setWordId(wordsRealm.get(i).getId());
                chunks.add(chunk);
            }
        }
        int chunksSize = chunks.size();
        int[] shuffledArray = RealmUtils.shuffleArray(chunksSize);
        for (int i = 0; i < chunksSize / 2; i++) {
            chunks.get(shuffledArray[i]).setPosition(shuffledArray[chunksSize - i - 1]);
            chunks.get(shuffledArray[chunksSize - i - 1]).setPosition(shuffledArray[i]);
        }
        return chunks;
    }

    public static List<Chunk> findSelectedChunksByLevelIdSorted(Realm realm, String levelId) {
        return realm
                .where(Chunk.class)
                .equalTo(REALM_FIELD_LEVEL_ID, levelId)
                .greaterThan(REALM_FIELD_STATE, CHUNK_STATE_NORMAL)
                .findAllSorted(REALM_FIELD_STATE);
    }

    public static RealmResults<Chunk> findSelectedChunksByLevelId(Realm realm, String levelId) {
        return realm
                .where(Chunk.class)
                .equalTo(REALM_FIELD_LEVEL_ID, levelId)
                .greaterThan(REALM_FIELD_STATE, CHUNK_STATE_NORMAL)
                .findAll();
    }

}

File: app/src/main/java/com/appchamp/wordchunks/data/ChunksRealmHelper.kt
package com.appchamp.wordchunks.data

import com.appchamp.wordchunks.models.realm.Chunk
import com.appchamp.wordchunks.models.realm.Word
import com.appchamp.wordchunks.util.Constants.*
import com.appchamp.wordchunks.util.shuffleIntArray
import io.realm.Realm
import io.realm.RealmList
import io.realm.RealmResults


/**
 * Chunks
 */
object ChunksRealmHelper {

    /**
     * Returns the list of Realm Chunk objects created from the array of split chunks.
     */
    internal fun createChunks(realm: Realm, wordsSplit: List<String>,
                              levelId: String, wordsRealm: RealmList<Word>): RealmList<Chunk> {
        val chunks = RealmList<Chunk>()
        for (i in wordsSplit.indices) {
            val splitChunks = wordsSplit[i].split(CHUNKS_SEPARATOR)
            for (chunkStr in splitChunks) {
                val chunk = realm.createObject(Chunk::class.java)
                chunk.chunk = chunkStr
                chunk.levelId = levelId
                chunk.wordId = wordsRealm[i].id
                chunks.add(chunk)
            }
        }
        val chunksSize = chunks.size
        val shuffledArray = IntArray(chunksSize, { i -> (i) }).shuffleIntArray()
        val size = if (chunksSize % 2 == 0) chunksSize / 2 else chunksSize / 2 + 1

        for (i in 0..size - 1) {
            chunks[shuffledArray[i]].position = shuffledArray[chunksSize - i - 1]
            chunks[shuffledArray[chunksSize - i - 1]].position = shuffledArray[i]
        }
        return chunks
    }

    fun findSelectedChunksByLevelIdSorted(realm: Realm, levelId: String): List<Chunk> = realm
            .where(Chunk::class.java)
            .equalTo(REALM_FIELD_LEVEL_ID, levelId)
            .greaterThan(REALM_FIELD_STATE, CHUNK_STATE_NORMAL)
            .findAllSorted(REALM_FIELD_STATE)

    fun findSelectedChunksByLevelId(realm: Realm, levelId: String): RealmResults<Chunk> = realm
            .where(Chunk::class.java)
            .equalTo(REALM_FIELD_LEVEL_ID, levelId)
            .greaterThan(REALM_FIELD_STATE, CHUNK_STATE_NORMAL)
            .findAll()
}